gravatar/src/main/java/com/gravatar/services/AvatarService.kt

```diff
@@ -48,4 +51,76 @@ public class AvatarService(okHttpClient: OkHttpClient? = null) {
             }
         }
     }
+
+    /**
+     * Uploads an image to be used as Gravatar avatar.
+     *
+     * @param file The image file to upload
+     * @param oauthToken The OAuth token to use for authentication
+     */
+    public suspend fun upload(file: File, oauthToken: String): Unit = withContext(GravatarSdkDI.dispatcherIO) {
+        val service = instance.getGravatarV3Service(okHttpClient, oauthToken)
+
+        val filePart =
+            MultipartBody.Part.createFormData("filedata", file.name, file.asRequestBody())
+
+        val response = service.uploadAvatar(filePart)
+
+        if (response.isSuccessful) {
+            Unit
+        } else {
+            // Log the response body for debugging purposes if the response is not successful
+            Logger.w(
+                LOG_TAG,
+                "Network call unsuccessful trying to upload Gravatar: $response.body",
+            )
+            throw HttpException(response)
+        }
+    }
+
+    /**
+     * Uploads an image to be used as Gravatar avatar.
+     * This method will catch any exception that occurs during the execution and return it as a [Result.Failure].
+     *
+     * @param file The image file to upload
+     * @param oauthToken The OAuth token to use for authentication
+     * @return The result of the operation
+     */
+    public suspend fun uploadCatching(file: File, oauthToken: String): Result<Unit, ErrorType> = runCatchingRequest {
+        upload(file, oauthToken)
+    }
+
+    /**
+     * Retrieves a list of available avatars for the authenticated user.
+     *
+     * @param oauthToken The OAuth token to use for authentication
+     * @return The list of avatars
+     */
+    public suspend fun retrieve(oauthToken: String): List<Avatar> = withContext(GravatarSdkDI.dispatcherIO) {
+        val service = instance.getGravatarV3Service(okHttpClient, oauthToken)
```

Comment:
❓ Creating these services on each function call seems unnecessary. Can't we use [Dynamic Headers](https://futurestud.io/tutorials/retrofit-add-custom-request-header) in Retrofit?

Reply:
That was my first idea. The problem was that API classes are generated by the OpenApi generator, and I haven't found a way for the generator to add those as a Dynamic Header. 

The only way I've found so far is to declare that header in the OpenApi specification explicitly. However, I don't like that solution, as we've already declared the security scheme the endpoint should use. It looks like a concrete problem with Kotlin/our generator, which should be solved without changes in the specifications.

I dedicated some time to this yesterday and finally decided to move on with this approach. As it's something internal in the library, we can try to find a way without retro compatibility issues.